docs/ENVIRONMENT.md
## Environment Variables

This environment variables can be used to configure released docker container at start time.
Also sample `.env` can be used as payload for `docker run` cli.

### General

| VAR_NAME              | Default Value               | Description |
| --------------------- | --------------------------- | ----------- |
| ERLANG_COOKIE         | `mFd1nWJ1KUxj`..            | Erlang [distribution cookie](http://erlang.org/doc/reference_manual/distributed.html). **Make sure that default value is changed in production.** |
| LOG_LEVEL             | `info`                      | Elixir Logger severity level. Possible values: `debug`, `info`, `warn`, `error`. |
| RECAPTCHA_VERIFY_URL  | `https://www.google.com...` | ReCAPTCHA verify URL|

### Database

Currently there are two options to set DB connections settings.

**By setting a separate value for each option:**

| VAR_NAME      | Default Value | Description |
| ------------- | ------------- | ----------- |
| DB_NAME       | `mithril_api_dev` | Database name. |
| DB_USER       | `postgres`    | Database user. |
| DB_PASSWORD   | `postgres`    | Database password. |
| DB_HOST       | `travis`      | Database host. |
| DB_PORT       | `5432`        | Database port. |

**By setting a single connection URI:**

| VAR_NAME      | Default Value | Description |
| ------------- | ------------- | ----------- |
| DATABASE_URL  | not set       | Database URI. Example: `postgres://postgres:postgres@travis:5432/mithril_api_dev` |

If you want to run migrations when container starts, use a `DB_MIGRATE` variable:

| VAR_NAME      | Default Value | Description |
| ------------- | ------------- | ----------- |
| DB_MIGRATE    | `true`        | Migrate database when container starts. |

### Phoenix HTTP Endpoint

| VAR_NAME      | Default Value | Description |
| ------------- | ------------- | ----------- |
| PORT          | `4000`        | HTTP host for web app to listen on. |
| HOST          | `localhost`   | HTTP port for web app to listen on. |
| SECRET_KEY    | `mFd1nWJ1KUxj`.. | Phoenix [`:secret_key_base`](https://hexdocs.pm/phoenix/Phoenix.Endpoint.html). **Make sure that default value is changed in production.** |

## Token lifetime in seconds

| VAR_NAME                    | Default Value    | Description              |
| --------------------------- | ---------------- | ------------------------ |
| AUTH_CODE_GRANT_LIFETIME    | 300              | Lifetime of code grant   |
| AUTH_ACCESS_TOKEN_LIFETIME  | 2592000          | Lifetime of access token |
| AUTH_REFRESH_TOKEN_LIFETIME | 604800           | Lifetime of secret token |

## Two-factor authentication

| VAR_NAME                           | Default Value                   | Description              |
| ---------------------------------- | ------------------------------- | ------------------------ |
| USER_2FA_ENABLED                   | true                            | Create or not default auth factor for user|
| USER_LOGIN_ERROR_MAX               | 3                               | Max failed user login attempts until user will be blocked |
| USER_OTP_ERROR_MAX                 | 3                               | Max failed OTP verifications attempts until user will be blocked |
| OTP_LENGTH                         | 6                               | OTP code length |
| OTP_LIFETIME                       | 300                             | OTP lifetime in seconds |
| OTP_EXPIRATION_SCHEDULE            | */5 * * * *                     | Cron config for running OTP expiration tasks |
| OTP_MAX_ATTEMPTS                   | 3                               | Max attempts for verification OTP |
| OTP_SMS_TEMPLATE                   | "Код підтвердження: <otp.code>" | SMS template for verification OTP. <otp.code> will replaced to OTP code |
| TRUSTED_CLIENT_IDS                 | not set                         | Client ids, that allowed not to send scopes for app:authorize method. E.g. Cabinet client ID |
| SENSITIVE_DATA_IN_RESPONSE_ENABLED | false                           | Sensitive data in response for tests. E.g. sms code for auth |

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](Https://conventionalcommits.org) for commit guidelines.

<!-- changelog -->

## [3.3.3](https://github.com/edenlabllc/mithril.api/compare/3.3.3...3.3.3) (2019-1-23)





apps/mithril_api/config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :mithril_api, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:mithril_api, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#
# Or read environment variables in runtime (!) as:
#
#     :var_name, "${ENV_VAR_NAME}"

config :mithril_api,
  namespace: Mithril,
  ecto_repos: [Mithril.Repo],
  system_user: {:system, "EHEALTH_SYSTEM_USER", "4261eacf-8008-4e62-899f-de1e2f7065f0"},
  sensitive_data_in_response: {:system, :boolean, "SENSITIVE_DATA_IN_RESPONSE_ENABLED", false},
  trusted_clients: {:system, :list, "TRUSTED_CLIENT_IDS", []},
  api_resolvers: [
    sms: Mithril.API.SMS,
    mpi: Mithril.API.MPI,
    digital_signature: Mithril.API.Signature,
    recaptcha: Mithril.ReCAPTCHA
  ],
  token_ttl_after_expiration: {:system, :integer, "TOKEN_TTL_AFTER_EXPIRATION_DAYS", 30}

# Configure your database
config :mithril_api, Mithril.Repo,
  adapter: Ecto.Adapters.Postgres,
  database: {:system, "DB_NAME", "mithril_api_dev"},
  username: {:system, "DB_USER", "postgres"},
  password: {:system, "DB_PASSWORD", "postgres"},
  hostname: {:system, "DB_HOST", "localhost"},
  port: {:system, :integer, "DB_PORT", 5432},
  loggers: [{Ecto.LoggerJSON, :log, [:info]}]

config :mithril_api, Mithril.ReCAPTCHA,
  url: {:system, "RECAPTCHA_VERIFY_URL", "https://www.google.com/recaptcha/api/siteverify"},
  secret: {:system, "RECAPTCHA_SECRET"}

config :mithril_api, :generators,
  migration: false,
  binary_id: true,
  sample_binary_id: "11111111-1111-1111-1111-111111111111"

config :mithril_api, Mithril.TokenAPI.Deactivator, limit: 500

config :mithril_api, :password,
  expiration: {:system, :integer, "PASSWORD_EXPIRATION_DAYS", 90},
  max_failed_logins: {:system, :integer, "MAX_FAILED_LOGINS", 10},
  # minutes
  max_failed_logins_period: {:system, :integer, "MAX_FAILED_LOGINS_PERIOD", 10}

config :mithril_api, :"2fa",
  user_2fa_enabled?: {:system, :boolean, "USER_2FA_ENABLED", true},
  sms_enabled?: {:system, :boolean, "SMS_ENABLED", true},
  # minutes
  otp_send_timeout: {:system, :integer, "OTP_SEND_TIMEOUT", 1},
  otp_send_counter_max: {:system, :integer, "OTP_SEND_COUNTER_MAX", 3},
  user_otp_error_max: {:system, :integer, "USER_OTP_ERROR_MAX", 3},
  # seconds
  otp_ttl: {:system, :integer, "OTP_LIFETIME", 300},
  otp_length: {:system, :integer, "OTP_LENGTH", 6},
  otp_max_attempts: {:system, :integer, "OTP_MAX_ATTEMPTS", 3},
  otp_sms_template: {:system, :string, "OTP_SMS_TEMPLATE", "Код підтвердження: <otp.code>"}

# Configures Guardian
config :mithril_api,
  jwt_secret: {:system, "JWT_SECRET"},
  ttl_login: {:system, :integer, "JWT_LOGIN_TTL"}

config :mithril_api, Mithril.Guardian,
  issuer: "EHealth",
  secret_key: {Confex, :fetch_env!, [:mithril_api, :jwt_secret]}

# Configures the endpoint
config :mithril_api, Mithril.Web.Endpoint,
  url: [host: "localhost"],
  secret_key_base: "6sOsW9uKv+8o8y/hIA3F4dNkJE2O35e2l6SaS9P/xW0+Nh9Fo59T6JHnl0GzBmio",
  render_errors: [view: EView.Views.PhoenixError, accepts: ~w(json)]

# Configures Elixir's Logger
config :logger, :console,
  format: "$message\n",
  handle_otp_reports: true,
  level: :info

config :mithril_api, :token_lifetime, %{
  code: {:system, :integer, "AUTH_CODE_GRANT_LIFETIME", 5 * 60},
  access: {:system, :integer, "AUTH_ACCESS_TOKEN_LIFETIME", 30 * 24 * 60 * 60},
  refresh: {:system, :integer, "AUTH_REFRESH_TOKEN_LIFETIME", 7 * 24 * 60 * 60}
}

# Configures OTP Verification API
config :mithril_api, Mithril.API.SMS,
  endpoint: {:system, "OTP_ENDPOINT"},
  hackney_options: [
    connect_timeout: {:system, :integer, "OTP_REQUEST_TIMEOUT", 30_000},
    recv_timeout: {:system, :integer, "OTP_REQUEST_TIMEOUT", 30_000},
    timeout: {:system, :integer, "OTP_REQUEST_TIMEOUT", 30_000}
  ]

# Configures MPI API
config :mithril_api, Mithril.API.MPI,
  endpoint: {:system, "MPI_ENDPOINT"},
  hackney_options: [
    connect_timeout: {:system, :integer, "MPI_REQUEST_TIMEOUT", 30_000},
    recv_timeout: {:system, :integer, "MPI_REQUEST_TIMEOUT", 30_000},
    timeout: {:system, :integer, "MPI_REQUEST_TIMEOUT", 30_000}
  ]

# Configures Digital Signature API
config :mithril_api, Mithril.API.Signature,
  enabled: {:system, :boolean, "DIGITAL_SIGNATURE_ENABLED", false},
  endpoint: {:system, "DIGITAL_SIGNATURE_ENDPOINT"},
  hackney_options: [
    connect_timeout: {:system, :integer, "DIGITAL_SIGNATURE_REQUEST_TIMEOUT", 30_000},
    recv_timeout: {:system, :integer, "DIGITAL_SIGNATURE_REQUEST_TIMEOUT", 30_000},
    timeout: {:system, :integer, "DIGITAL_SIGNATURE_REQUEST_TIMEOUT", 30_000}
  ]

config :mithril_api, Mithril.Scheduler,
  token_expiration: {:system, :string, "TOKEN_EXPIRATION_SCHEDULE", "* 0-4 * * *"},
  token_deleting: {:system, :string, "TOKEN_DELETING_SCHEDULE", "* 1-4 * * *"},
  otp_expiration: {:system, :string, "OTP_EXPIRATION_SCHEDULE", "*/5 * * * *"}

import_config "#{Mix.env()}.exs"
